File: elgcsharp/KeyPairGenerator.cs
namespace sl.crypto.elgamal.generators
{	
	using sl.crypto.elgamal.keys;	
	using sl.crypto.elgamal.parameter;
	using sl.crypto.util;	
	
	using System;

	/**
	 * <summary>
 	 * Autor: Matthias Koch
 	 * Version: 0.5
 	 * 
 	 * Klasse welche ein Objekt zur erzeugung eines Schlüsselpaares repräsentiert.
 	 * </summary> 
	 */
	public class KeyPairGenerator
	{
		private PublicKey publicKey;
		// erlaube lesenden Zugriff
		public PublicKey PublicKey
		{
			get{return publicKey;}
		}
		private PrivateKey privateKey;
		//erlaube lesenden Zugrriff
		public PrivateKey PrivateKey
		{
			get{return privateKey;}
		}
	
		private Parameter parameter;
		private Random random;
		private int strength;
	
		/**
		 * 
		 * <param name="random"/>
		 * <param name="params"/>
		 */	
		public KeyPairGenerator(Random random, Parameter parameter)
		{
			this.parameter = parameter;
			this.random = random;
			this.strength = parameter.P.bitCount() - 1;			
			this.generateKeyPair();
		}
		/**
		 *
		 */
		private void generateKeyPair()
		{
			BigInteger p, g, x, y;			
			int qLength = strength - 1;
			p = parameter.P;			
			g = parameter.G;
			//
			// berechne den private key
			//
			x = new BigInteger();
			x.genRandomBits(qLength, random);
			//
			// berechne den public key.
			//
			y = g.modPow(x, p);
			this.publicKey = new PublicKey(y, parameter);
			this.privateKey = new PrivateKey(x, parameter);			
		}
	}
}

File: elgjava/sl/crypto/elgamal/generators/KeyPairGenerator.java
package sl.crypto.elgamal.generators;

import java.math.BigInteger;
import java.security.SecureRandom;

import sl.crypto.elgamal.keys.PrivateKey;
import sl.crypto.elgamal.keys.PublicKey;
import sl.crypto.elgamal.parameter.Parameter;
/**
 * Klasse welche ein Objekt zur erzeugung eines Schlüsselpaares repräsentiert.
 * 
 * @author Matthias Koch
 * @version 0.5   
 */
public class KeyPairGenerator
{
	private PublicKey publicKey;
	private PrivateKey privateKey;
	
	private Parameter params;
	private SecureRandom random;
	private int strength;
	/**
	 * 
	 * @param random
	 * @param params
	 */	
	public KeyPairGenerator(SecureRandom random, Parameter params)
	{
		this.params = params;
		this.random = random;
		this.strength = params.getP().bitLength() - 1;
		this.generateKeyPair();
	}
	/**
	 * gibt den öffentlichen Schlüssel zurück
	 *
	 * @return the public key parameters.
	 */
	public PublicKey getPublic()
	{
		return publicKey;
	}
	/**
	 * gibt den privaten Schlüssel zurück
	 *
	 * @return the private key parameters.
	 */
	public PrivateKey getPrivate()
	{
		return privateKey;
	}

	/**
	 * 
	 */
	private void generateKeyPair()
	{
		BigInteger p, g, x, y;
		int qLength = this.strength - 1;
		p = params.getP();
		g = params.getG();
		//
		// berechne den private key
		//		
		x = new BigInteger(qLength, this.random);
		//
		// berechnde den public key.
		//
		y = g.modPow(x, p);
		this.publicKey = new PublicKey(y, params);
		this.privateKey = new PrivateKey(x, params);		
	}
}
